okonClient.py

```diff
@@ -0,0 +1,349 @@
+        self.okon = Okon(self)
+        self.simulation = Simulation(self)
+
+        self.ip = ip
+        self.port = port
+        self.debug = debug
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.connected = False
+        self._events = dict()
+        self.syncTime = time.time()
+        self.sync_interval = sync_interval
+        self._to_send = queue.Queue()
+    
+    def connect(self) -> bool:
+        if self.debug:
+            print(f'Connecting to {self.ip}:{self.port}')
+        try:
+            self.socket.connect((self.ip, self.port))
+            self.connected = True
+            start_new_thread(self._comm_thread, ())
+            start_new_thread(self._sync_thread, ())
+            return True
+        except Exception as err:
+            print(f'failed to connect {err}')
+            return False
+
+    def send(self, packet_type : int, packet_flag : int= PacketFlag.NONE, json : str = None) -> None:
+        if json == None:
+            self._to_send.put((packet_type, packet_flag, 0))
+        else:
+            json_bytes = json.encode()   
+            self._to_send.put((packet_type, packet_flag, len(json_bytes), json_bytes))     
+
+    def _sync_thread(self) -> None:
+        while self.connected:
+            while not self._to_send.empty():
+                packet = self._to_send.get()
+                self._send(packet)
+
+            if time.time() > self.syncTime + self.sync_interval:
+                self.syncTime = time.time()
+                self.simulation.sync()
+            else:
+                time.sleep(0.001) # avg ping ~2.7ms
+
+    def _send(self, packet : tuple) -> None:
+        self.socket.sendall(packet[0].to_bytes(1, byteorder='little'))
+        self.socket.sendall(packet[1].to_bytes(1, byteorder='little'))
+        self.socket.sendall(packet[2].to_bytes(4, byteorder='little'))
+        if packet[2] > 0:
+            self.socket.sendall(packet[3])
+            if self.debug and packet[1] & PacketFlag.DO_NOT_LOG_PACKET == 0:
+                print(f'SENT {PacketType.get(packet[0])} {PacketFlag.get(packet[1])} dat:{packet[3].decode("utf-8") }') 
+        else:
+            if self.debug and packet[1] & PacketFlag.DO_NOT_LOG_PACKET == 0:
+                print(f'SENT {PacketType.get(packet[0])} {PacketFlag.get(packet[1])}') 
+    
+    def _receiveAll(self, n : int) -> bytes:
+        buffer = b''
+        while len(buffer) != n:
+            buffer += self.socket.recv(n-len(buffer))
+        return buffer
+    
+    def _comm_thread(self) -> None:
+        while self.connected:
+            packet_typeFlag = self._receiveAll(2)
+            dataLength = int.from_bytes(self._receiveAll(4),"little")
+            data_bytes = self._receiveAll(dataLength)
+            self._handle_packet(packet_typeFlag[0], packet_typeFlag[1], data_bytes)
+    
+    def _handle_packet(self, packet_type : int, packet_flag : int, data_bytes : bytes):
```

Comment:
About this huge `if` block. In my opinion the solution could making a dict mapping PacketFlag.value to assigned function, just like `[here](https://www.geeksforgeeks.org/python-store-function-as-dictionary-value/)`. You can even call functions with parameters.
As said before, you would need to prepare a dictionary with reverse relation to `vars(PacketType)`. You could do this like that:
```
class Testy:
  def __init__(self):
    self.a = 1 # you could do those byte values 
    self.b = 2
    self.c = 3

  make_standard(self):
    return vars(self)
  def make_reverse(self): 
    return  {v:k for k, v in vars(self).items()} #funny dict comprehension

# now interpreter part
>>> testy = Testy()
>>> vars(Testy)
mappingproxy({'__module__': '__main__', '__init__': <function Testy.__init__ at 0x00000277DA06C3A0>, 'make_standard': <function Testy.make_standard at 0x00000277DA06C430>, 'make_reverse': <function Testy.make_reverse at 0x00000277DA06C4C0>, '__dict__': <attribute '__dict__' of 'Testy' objects>, '__weakref__': <attribute '__weakref__' of 'Testy' objects>, '__doc__': None
>>> vars(testy)
{'a': 1, 'b': 2, 'c': 3}                                                                                                
>>> testy.make_standard()
{'a': 1, 'b': 2, 'c': 3}
>>> testy.make_reverse()
{1: 'a', 2: 'b', 3: 'c'}
>>> testy.reverse[testy.standard["a"]]
"a"
```
Python can use bytes as keys in dictionary. so don't worry about some problems in reversing of this dictionary. As you can see I use constructor and I generally prefere working on instances rather on classes (like you do). If you use vars on whole class, you will get few unnecessary fields. I proposed dataclsses before, because they get init done for you (much cleaner code), but you had problems in serializing them to json.

When you have such mapping, you could just prepare another dict with matching functions to each PacketType.xxx.
```
class Mapper:
  def __init__(self, testy):
    self.standard = testy.make_standard()
    self.reverse = testy.make_reverse()
   def make_methods(self):
     def method_a():
        print("hello, method_a")
      def method_b():
        print("hello, method_b")
      def method_c():
        print("hello, method_c")
      self.methods = {"a": method_a, "b": method_b, "c": method_c}
  def mapping(self, tested_val):
    self.methods[self.reverse[tested_val]]()

# interpreter part
>>> mapper = Mapper(testy)
>>> mapper.make_methods()
>>> mapper.methods["a"]                                                                                                                                                                                                                      <function Mapper.make_methods.<locals>.method_a at 0x00000277D9B4DF70>
>>> mapper.mapping(testy.a)
hello, method_a!
>>> for k in list(testy.make_reverse()):
...  mapper.mapping(k)
...
hello, method_a!
hello, method_b!
hello, method_c!
```
You could add parametrization to these functions, just adding the parameters in the parenthesis in mapping function code (also you should get those parameters. If parameters are flag-specific you could make `methods` dict of tuples `(function, parameters)`

Reply:
Also I'm pretty sure you could automate this `methods` dict making with dict comprehension -  you already know the fields of `Testy` and you make methods with names corresponding to those fields, so it feels like you could automate this somehow 😄

Reply:
Did some benchmarking and:
```
type -> number dynamic class dict   0.4790667
type -> number static class dict    0.4685557
type -> number class dynamic        0.4750073
type -> number class static         0.4473778
number -> type dynamic class dict   0.5743588
number -> type static class dict    0.4793005
number -> type elif                 8.4772784
```
Best solution is to keep static class based (PacketType.XXX) and add inverse dict to replace huge if block. Anything is better than elif >.<
(if tested in the worse case going through 30 ifs)